### PR TITLE
Add p256 envelope encryption

### PR DESCRIPTION
**Pull request by emostov**
Add p256 envelope encryption.

The plan is to have:

- abstraction over encryption key pair (P256EncryptionPair) <-- this pr
- abstraction over authentication key pair (P256AuthenticationPair) <-- follow up
- abstraction over authentication and encryption key pairs that utilizes the aforementioned (P256Pair). For example, the QuorumKey will fall under this as it will be composite of an authentication secret and a separate encryption secret. This abstraction should make development simpler as the secret will always be stored and handled together. <-- follow up

### CODE REVIEW

File: src/qos_p256/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	FailedToCreateAes256GcmCipher,
+	/// The public key could not be deserialized.
+	FailedToDeserializePublicKey,
+}
+
+/// Envelope for serializing an encrypted message with it's context.
+#[derive(Debug, borsh::BorshSerialize, borsh::BorshDeserialize)]
+struct Envelope {
+	/// Nonce used as an input to the cipher.
+	nonce: Vec<u8>,
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]:
```suggestion
	nonce: [u8; 12],
```

Comment (mikelodder7) [reply src/qos_p256/src/lib.rs]: Having a fixed length can help prevent denial of service attacks. An attacker could supply a GB sized nonce to your payload and crash it.

File: src/qos_p256/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	/// Public key as sec1 encoded point with no compression
+	ephemeral_public: Vec<u8>,
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]: Same here try to have a fixed size because it won't ever change.

File: src/qos_p256/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		// doesn't seem strictly necessary
+		let encrypted_message = cipher
+			.encrypt(&nonce, message)
+			.map_err(|_| P256Error::AesGcm256EncryptError)?;
+
+		let envelope = Envelope {
+			encrypted_message,
+			nonce: nonce.to_vec(),
+			ephemeral_public: ephemeral_public
+				// TODO: Should we do compression? Is there a better way to
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]: Compression is not preferred for P256 keys unfortunately according to NIST.

File: src/qos_p256/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	// TODO: from der/sec1 etc
+}
+
+// Helper function to create the `Aes256Gcm` cypher.
+fn create_cipher(
+	private: &EphemeralSecret,
+	public: &PublicKey,
+) -> Result<Aes256Gcm, P256Error> {
+	let shared_secret = private.diffie_hellman(public);
+	let shared_key = sha2::Sha512::digest(shared_secret.raw_secret_bytes());
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]: Hash the public keys in with it. It helps add entropy even though they are known plus it adds a domain context.
So hash it like so `Sha512(SenderPK || ReceiverPK || shared_secret)`

File: src/qos_p256/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+			.map_err(|_| P256Error::AesGcm256EncryptError)?;
+
+		let envelope = Envelope {
+			encrypted_message,
+			nonce: nonce.to_vec(),
+			ephemeral_public: ephemeral_public
+				// TODO: Should we do compression? Is there a better way to
+				// serialize the public key.
+				.to_encoded_point(false)
+				.as_ref()
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]: Use fixed size array instead here to prevent oversized payload key deserialization attacks.

File: src/qos_p256/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	/// Nonce used as an input to the cipher.
+	nonce: Vec<u8>,
+	/// Public key as sec1 encoded point with no compression
+	ephemeral_public: Vec<u8>,
+	/// The data encrypted with an AES 256 GCM cipher.
+	encrypted_message: Vec<u8>,
+}
+
+/// P256 key pair
+pub struct P256Pair {
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]:
```suggestion
#[transparent]
pub struct P256Pair(EphemeralSecret);
```

Comment (mikelodder7) [reply src/qos_p256/src/lib.rs]: Sometimes helps with cleanup and prevents copies in memory but you'll have to test this with something like vagrant.

File: src/qos_p256/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// Helper function to create the `Aes256Gcm` cypher.
+fn create_cipher(
+	private: &EphemeralSecret,
+	public: &PublicKey,
+) -> Result<Aes256Gcm, P256Error> {
+	let shared_secret = private.diffie_hellman(public);
+	let shared_key = sha2::Sha512::digest(shared_secret.raw_secret_bytes());
+	Aes256Gcm::new_from_slice(&shared_key[..AES256_KEY_LEN])
```
Comment (mikelodder7) [inline src/qos_p256/src/lib.rs]: I don't see any AAD data added. Use the `SenderPK || ReceiverPK || nonce`

File: src/qos_p256/src/lib.rs (hunk 1 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: ef7e4a5ca7c0e561c09a83133996ae75f29fe628

File: src/qos_p256/src/lib.rs (hunk 2 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: ef7e4a5ca7c0e561c09a83133996ae75f29fe628

File: src/qos_p256/src/lib.rs (hunk 5 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: ef7e4a5ca7c0e561c09a83133996ae75f29fe628

File: src/qos_p256/src/lib.rs (hunk 6 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: Passing on this for now because we cannot use nightly with our deterministic build tooling

File: src/qos_p256/src/lib.rs (hunk 7 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: [46661d5](https://github.com/tkhq/qos/pull/150/commits/46661d5049087be5e459babdebd3e661d89f0f07)

File: src/qos_p256/src/lib.rs (hunk 4 above)
Comment (emostov) [reply src/qos_p256/src/lib.rs]: [46661d5](https://github.com/tkhq/qos/pull/150/commits/46661d5049087be5e459babdebd3e661d89f0f07)

File: src/qos_p256/src/encrypt.rs [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+
+// Helper function to create the `Aes256Gcm` cypher.
+fn create_cipher(
+	private: &EphemeralSecret,
+	public: &PublicKey,
+	ephemeral_sender_public: &SenderPublic,
+	receiver_public: &ReceiverPublic,
+) -> Result<Aes256Gcm, P256Error> {
+	let shared_secret = private.diffie_hellman(public);
+	//  To help with entropy and add domain context, we do
```
Comment (emostov) [inline src/qos_p256/src/encrypt.rs]:
```suggestion
	// To help with entropy and add domain context, we do
```

File: src/qos_p256/src/encrypt.rs [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+		let aad = create_additional_associated_data(
+			&sender_public_typed,
+			&receiver_public_typed,
+			nonce.as_ref(),
+		);
+		let payload = Payload { aad: &aad, msg: message };
+
+		// TODO: use nonce||ephemeral_public as authenticated data .. although
+		// doesn't seem strictly necessary
```
Comment (r-n-o) [inline src/qos_p256/src/encrypt.rs]: Can you open an issue if we want to tackle this later? Then let's reference it, e.g.:
```rust
// TODO(#ISSUE_NUM): use nonce||ephemeral_public as authenticated data
```

File: src/qos_p256/src/encrypt.rs [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+		let nonce = nonce
+			.try_into()
+			.map_err(|_| P256Error::FailedToCoerceNonceToIntendedLength)?;
+
+		let envelope =
+			Envelope { nonce, ephemeral_sender_public, encrypted_message };
+
+		envelope.try_to_vec().map_err(|_| P256Error::FailedToSerializeEnvelope)
+	}
+	// TODO: from der/sec1 etc
```
Comment (r-n-o) [inline src/qos_p256/src/encrypt.rs]: What do you mean by this?

File: src/qos_p256/src/encrypt.rs [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+		// byte and shift everthing after, making the nonce too long.
+		serialized_envelope.insert(BITS_96_AS_BYTES, 0xff);
+
+		assert_eq!(
+			alice_pair.decrypt(&serialized_envelope).unwrap_err(),
+			P256Error::FailedToDeserializeEnvelope
+		);
+	}
+
+	// What other edge cases should we test for?
```
Comment (r-n-o) [inline src/qos_p256/src/encrypt.rs]: I think you can remove this comment

File: src/qos_p256/src/encrypt.rs (hunk 9 above)
Comment (emostov) [reply src/qos_p256/src/encrypt.rs]: Ah good catch. Already addressed above. See `create_additional_associated_data`

Comment (emostov) [reply src/qos_p256/src/encrypt.rs]: will remove

File: src/qos_p256/src/encrypt.rs (hunk 10 above)
Comment (emostov) [reply src/qos_p256/src/encrypt.rs]: Just we need to expose serialization functions. Will remove
